test/node-fetch/main.js

```diff
@@ -0,0 +1,2310 @@
+  //     method: 'PATCH',
+  //     body: 'a=1'
+  //   }
+  //   return fetch(url, options).then(res => {
+  //     expect(res.url).to.equal(`${base}inspect`)
+  //     expect(res.status).to.equal(200)
+  //     return res.json().then(result => {
+  //       expect(result.method).to.equal('PATCH')
+  //       expect(result.body).to.equal('a=1')
+  //     })
+  //   })
+  // })
+
+  // it('should not follow non-GET redirect if body is a readable stream', () => {
+  //   const url = `${base}redirect/307`
+  //   const options = {
+  //     method: 'PATCH',
+  //     body: stream.Readable.from('tada')
+  //   }
+  //   return expect(fetch(url, options)).to.eventually.be.rejected
+  //     .and.be.an.instanceOf(FetchError)
+  //     .and.have.property('type', 'unsupported-redirect')
+  // })
+
+  // it('should obey maximum redirect, reject case', () => {
+  //   const url = `${base}redirect/chain`
+  //   const options = {
+  //     follow: 1
+  //   }
+  //   return expect(fetch(url, options)).to.eventually.be.rejected
+  //     .and.be.an.instanceOf(FetchError)
+  //     .and.have.property('type', 'max-redirect')
+  // })
+
+  // it('should obey redirect chain, resolve case', () => {
+  //   const url = `${base}redirect/chain`
+  //   const options = {
+  //     follow: 2
+  //   }
+  //   return fetch(url, options).then(res => {
+  //     expect(res.url).to.equal(`${base}inspect`)
+  //     expect(res.status).to.equal(200)
+  //   })
+  // })
+
+  // it('should allow not following redirect', () => {
+  //   const url = `${base}redirect/301`
+  //   const options = {
+  //     follow: 0
+  //   }
+  //   return expect(fetch(url, options)).to.eventually.be.rejected
+  //     .and.be.an.instanceOf(FetchError)
+  //     .and.have.property('type', 'max-redirect')
+  // })
+
+  // it('should support redirect mode, manual flag', () => {
+  //   const url = `${base}redirect/301`
+  //   const options = {
+  //     redirect: 'manual'
+  //   }
+  //   return fetch(url, options).then(res => {
+  //     expect(res.url).to.equal(url)
+  //     expect(res.status).to.equal(301)
+  //     expect(res.headers.get('location')).to.equal(`${base}inspect`)
+  //   })
+  // })
+
+  // it('should support redirect mode, manual flag, broken Location header', () => {
+  //   const url = `${base}redirect/bad-location`
+  //   const options = {
+  //     redirect: 'manual'
+  //   }
+  //   return fetch(url, options).then(res => {
+  //     expect(res.url).to.equal(url)
+  //     expect(res.status).to.equal(301)
+  //     expect(res.headers.get('location')).to.equal(`${base}redirect/%C3%A2%C2%98%C2%83`)
+  //   })
+  // })
+
+  // it('should support redirect mode, error flag', () => {
+  //   const url = `${base}redirect/301`
+  //   const options = {
+  //     redirect: 'error'
+  //   }
+  //   return expect(fetch(url, options)).to.eventually.be.rejected
+  //     .and.be.an.instanceOf(FetchError)
+  //     .and.have.property('type', 'no-redirect')
+  // })
```

Comment:
@szmarczak: Do you want to take a look at the redirect failures?

Reply:
sure

Reply:
`FetchError` doesn't exist. https://fetch.spec.whatwg.org/#concept-request-redirect-mode

> "error"
> Return a network error when a request is met with a redirect.

> A response whose type is "error" is known as a network error.
> 
> A network error is a response whose status is always 0, status message is always the empty byte sequence, header list is always empty, and body is always null.

I'll look how it's done in browsers.

Reply:
You can just comment out `FetchError` or replace it with `Error`. Please try all the redirect tests.

Reply:
Hmm... Google Chrome throws `TypeError: Failed to fetch`. Firefox:

```
TypeError: NetworkError when attempting to fetch resource.
```

Reply:
I think we could just create this empty response and reject the promise with it

Reply:
> I think we could just create this empty response and reject the promise with it

Ah I see it's done already, let me fix the tests then :)

Reply:
I'm having a hard time communicating with `chai`:

```js
  it.only('should support redirect mode, error flag', async () => {
    const url = `${base}redirect/301`
    const options = {
      redirect: 'error'
    }

    return expect(fetch(url, options)).to.be.fulfilled
      .and.be.instanceOf(Response).and.have.property('status', 0)
  })
```

should pass but gives

```js
AssertionError: expected {} to be an instance of Respons
```

btw

```js
console.log((await fetch(url, options)) instanceof Response) //=> true
```

Reply:
Ok this works (`chai` requires `.eventually` for promises):

```js
  it('should support redirect mode, error flag', async () => {
    const url = `${base}redirect/301`
    const options = {
      redirect: 'error'
    }

    return expect(fetch(url, options)).to.be.fulfilled
      .and.eventually.be.instanceOf(Response).and.have.property('status', 0)
  })
```

Reply:
This changes the node fetch test though? It thinks we should reject promise here. Is that a spec violation?

Reply:
I don't think redirects are working properly at all atm. Hopefully someone can take a look at that.

Reply:
It does violate the spec. https://fetch.spec.whatwg.org/#fetch-method

> If response is a network error, then reject p with a TypeError and terminate these substeps.